libdnf5/rpm/solv/goal_private.cpp

```diff
@@ -262,61 +268,55 @@ bool GoalPrivate::limit_installonly_packages(libdnf5::solv::IdQueue & job, Id ru
+            available_unused_providers.push_back(solvable);
         }
-        if (!installing.size()) {
+    }
+
+    const InstallonlyCmpData installonly_cmp_data{spool, running_kernel};
+    marked_for_install_providers.sort(&installonly_cmp, &installonly_cmp_data);
+    std::sort(available_unused_providers.begin(), available_unused_providers.end(), name_solvable_cmp_key);
+
+    libdnf5::solv::IdQueue same_names;
+    while (!marked_for_install_providers.empty()) {
+        Id name = same_name_subqueue(spool, &marked_for_install_providers.get_queue(), &same_names.get_queue());
+        if (same_names.size() <= static_cast<int>(installonly_limit)) {
             continue;
         }
 
-        const InstallonlyCmpData installonly_cmp_data{spool, running_kernel};
-        q.sort(&installonly_cmp, &installonly_cmp_data);
-        std::sort(available_unused_providers.begin(), available_unused_providers.end(), name_solvable_cmp_key);
+        // We want to avoid reinstalling packages marked for ERASE, therefore
+        // if some unused provider is also available we need to mark it ERASE as well.
+        auto low = std::lower_bound(
+            available_unused_providers.begin(), available_unused_providers.end(), name, name_solvable_id_name_cmp_key);
+        while (low != available_unused_providers.end() && (*low)->name == name) {
+            job.push_back(SOLVER_ERASE | SOLVER_SOLVABLE, spool.solvable2id(*low));
+            ++low;
+        }
 
-        libdnf5::solv::IdQueue same_names;
-        while (q.size() > 0) {
-            same_name_subqueue(spool, &q.get_queue(), &same_names.get_queue());
-            if (same_names.size() <= static_cast<int>(installonly_limit)) {
-                continue;
-            }
-            reresolve = true;
-            for (int j = 0; j < same_names.size(); ++j) {
-                Id id = same_names[j];
-                Id action = SOLVER_ERASE;
-                if (j < static_cast<int>(installonly_limit)) {
-                    action = SOLVER_INSTALL;
-                } else {
-                    // We want to avoid reinstalling packages marked for ERASE, therefore
-                    // if some unused provider is also available we need to mark it ERASE as well.
-                    Solvable * solvable = spool.id2solvable(id);
-                    auto low = std::lower_bound(
-                        available_unused_providers.begin(),
-                        available_unused_providers.end(),
-                        solvable,
-                        nevra_solvable_cmp_key);
-                    while (low != available_unused_providers.end() && (*low)->name == solvable->name) {
-                        job.push_back(SOLVER_ERASE | SOLVER_SOLVABLE, spool.solvable2id(*low));
-                        ++low;
-                    }
-                }
-                job.push_back(action | SOLVER_SOLVABLE, id);
+        reresolve = true;
+        for (int j = 0; j < same_names.size(); ++j) {
```

Comment:
I believe that iterator might be modify to not use indexes. What do you think?

Reply:
The index `j` is also used as a counter for the `installonly_limit`. I find this approach simpler than having a separate counter variable so I would prefer to keep this.

Reply:
Good point that I overlooked. Please keep it like it is.